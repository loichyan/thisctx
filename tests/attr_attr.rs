use thisctx::WithContext;

#[derive(WithContext)]
#[thisctx(attr(doc = "I'm on Line#2\n\n"))]
#[thisctx(attr(derive(Clone)))]
pub enum Error {
    #[thisctx(attr(doc = "I'm on Line#1\n\n"))]
    #[thisctx(attr(derive(Copy)))]
    ExtendAttributes(String),
    #[thisctx(attr(doc = "I'm also on Line#1\n\n"))]
    FieldAttributes(#[thisctx(attr(doc = "I'm a field"))] String),
}

fn requires_copied<T: Copy>(_: T) {}

#[test]
fn attr() {
    let ctx = ExtendAttributesContext("What's going on?");
    requires_copied(ctx);
}
