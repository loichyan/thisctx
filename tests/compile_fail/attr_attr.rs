#![allow(dead_code)]

#[derive(thisctx::WithContext)]
#[thisctx(attr = "derive(Debug)")]
pub enum Error {
    DebugDerived(String),
    #[thisctx(attr = "cfg(all())")]
    DebugNotDerived(String),
    #[thisctx(attr = "derive(Clone, Copy)")]
    #[error(transparent)]
    CopyDerived(String),
}

fn requires_debug<T: std::fmt::Debug>(_: T) {}
fn requires_copied<T: Copy>(_: T) {}

fn main() {
    requires_debug(DebugDerived("with some generic magic"));
    requires_debug(DebugNotDerived("parent attributes are overridden"));
    requires_copied(CopyDerived);
}
