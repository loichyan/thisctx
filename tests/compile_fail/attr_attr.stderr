error[E0277]: `DebugNotDerived<&str>` doesn't implement `Debug`
  --> tests/compile_fail/attr_attr.rs:18:20
   |
18 |     requires_debug(DebugNotDerived("parent attributes are overridden"));
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `DebugNotDerived<&str>` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `DebugNotDerived<&str>`
   = note: add `#[derive(Debug)]` to `DebugNotDerived<&str>` or manually `impl Debug for DebugNotDerived<&str>`
note: required by a bound in `requires_debug`
  --> tests/compile_fail/attr_attr.rs:13:22
   |
13 | fn requires_debug<T: std::fmt::Debug>(_: T) {}
   |                      ^^^^^^^^^^^^^^^ required by this bound in `requires_debug`
